bytecomp/matching.ml

```diff
@@ -1178,160 +1193,66 @@ let rec split_or argo cls args def =
           (Default_environment.cons matrix idef def, (idef, next) :: nexts)
     in
     match yesor with
-    | [] -> split_constr yes args def nexts
+    | [] -> split_no_or yes args def nexts
     | _ -> precompile_or argo yes yesor args def nexts
   in
   do_split [] [] [] cls
 
-(* Ultra-naive splitting, close to semantics, used for extension,
-   as potential rebind prevents any kind of optimisation
-
-   Indeed, extension constructors with distinct names may be equal thanks to
-   constructor rebinding. This is compiled by having a specialized
-   submatrix for each syntactically-distinct constructor, with a threading
-   of exits such that each submatrix falls back to the potentially-compatible
-   submatrices below it.
-*)
-and split_naive cls args def k =
-  let rec split_exc cstr0 rev_yes = function
-    | [] ->
-        let yes = List.rev rev_yes in
-        ( { me = Pm { cases = yes; args; default = def };
-            matrix = as_matrix yes;
-            top_default = def
-          },
-          k )
-    | ((p :: _, _) as cl) :: rem ->
-        if group_constructor p then
-          let cstr = pat_as_constr p in
-          if cstr = cstr0 then
-            split_exc cstr0 (cl :: rev_yes) rem
-          else
-            let yes = List.rev rev_yes in
-            let { me = next; matrix; top_default = def }, nexts =
-              split_exc cstr [ cl ] rem
-            in
-            let idef = next_raise_count () in
-            let def = Default_environment.cons matrix idef def in
-            ( { me = Pm { cases = yes; args; default = def };
-                matrix = as_matrix yes;
-                top_default = def
-              },
-              (idef, next) :: nexts )
-        else
-          let yes = List.rev rev_yes in
-          let { me = next; matrix; top_default = def }, nexts =
-            split_noexc [ cl ] rem
-          in
-          let idef = next_raise_count () in
-          let def = Default_environment.cons matrix idef def in
-          ( { me = Pm { cases = yes; args; default = def };
-              matrix = as_matrix yes;
-              top_default = def
-            },
-            (idef, next) :: nexts )
-    | _ -> assert false
-  and split_noexc rev_yes = function
+and split_no_or cls args def k =
+  let rec split cls =
+    let discr = what_is_first_case cls in
+    collect discr [] [] cls
+  and collect group_discr rev_yes rev_no = function
     | ([], _) :: _ -> assert false
-    | [] -> precompile_var args (List.rev rev_yes) def k
-    | ((p :: _, _) as cl) :: rem ->
-        if not (group_constructor p) then
-          split_noexc (cl :: rev_yes) rem
+    | [ ((ps, _) as cl) ] when List.for_all group_var ps && rev_yes <> [] ->
+        (* This enables an extra division in some frequent cases:
+               last row is made of variables only *)
+        collect group_discr rev_yes (cl :: rev_no) []
```

Comment:
do we know anything about what this special case does in the non-variable case?

Reply:
(you could avoid the question by checking `group_var group_discr` as an extra condition)

Reply:
(@trefis replied in the main discussion that, in the non-variable case, this creates a different group anyway.)